### PR TITLE
feat(cosmo): `angular_diameter_distance(z1,z2)`

### PR DESCRIPTION
**Pull request by nstarman**
This PR upgrades `angular_diameter_distance` to be a two-argument form, deprecating the now unnecessary `angular_diameter_distance_z1z2`.  Previously this was done for `comoving_distance`, https://github.com/astropy/astropy/blob/19a688f7f87e5c3ab18608d67b3fa830b93690a0/astropy/cosmology/_src/flrw/base.py#L857-L863

~Requires #18800~ Complete!

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by nstarman** [comment]
@mhvk @neutrinoceros, friendly ping. hopefully an easy review :)

**Comment by nstarman** [comment]
Rebased while fixing a typo.

File: astropy/cosmology/_src/flrw/base.py [hunk 1]
```diff
@@ -962,17 +960,33 @@ def _comoving_transverse_distance_z1z2(
...
-        distance corresponding to an angle of 1 radian for an object
-        at redshift ``z`` ([1]_, [2]_, [3]_).
+        The two redshift form is useful for e.g. gravitational lensing for computing the
+        angular diameter distance between a lensed galaxy and the foreground lens.
 
         Parameters
         ----------
-        z : Quantity-like ['redshift'], array-like
-            Input redshift.
+        z1, z2 : Quantity-like ['redshift'], array-like
```
Comment (mhvk) [inline astropy/cosmology/_src/flrw/base.py:978]: Are you sure you want to arange (pun intented) things in this order? It reminds be a bit of `range` and `np.arange()`, where one argument is taken to mean `0, arg0`, while two arguments are used as `arg0, arg1`. @neutrinoceros recently spent much more time than he would have liked getting that right for `np.arange(..., like=u.Quantity)`...

Might it be better to have something like `angular_diameter_distance(z, /, *, z_obs=0.0)`? It means more adaptation for those coming from the `_z1z2` variant, but would seem much clearer. (Clear enough, in effect, that you could remove the warning for `z2 < z1`.)

**Review by mhvk — COMMENTED** [review]
Like getting these two together, but see suggestion about signature in-line. This would seem to be the time to have it make sense...

File: astropy/cosmology/_src/flrw/base.py (hunk 1 above)
Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:978]: I definitely don't want `z_obs` as a keyword argument as these aren't compatible with most jit operations, which I very much want to be able support as we adopt the array-api.

IMO `func(z)` and `func(z1, z2)` are close enough to how these functions are written mathematically (e.g. $d_A(z)$ and $d_A(z1,z2)$) that it's not confusing. Also, the observer doesn't have to be at `z_obs`, that's just where the calculation stops. In a ratio of distances, for example, $d_A(z1,z2)/$d_A(z_obs,z2)$ comparisons are being made relative to an observer.
IMO these should be positional-only arguments because it's the order that matters, not the variable name.

Comment (mhvk) [reply astropy/cosmology/_src/flrw/base.py:978]: It doesn't have to be keyword-only, but effectively having default positional arguments before anything else just does not feel right (for once, typing shows this, as it requires the overload). How about `*_distance(z, /, z_ref=0.0)` as an alternative? (The mathematical precedent is just as bad; z1 and z2 are just poor choices of names, at least use `z_lens`, `z_src` or something like it.)

Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:978]: > but effectively having default positional arguments before anything else just does not feel right

Maybe I'm not understanding. There isn't a default positional arguments before anything else. It proceeds `z1`.

> for once, typing shows this, as it requires the overload

It actually doesn't. 
Without the `overloads` it's:
```python
def angular_diameter_distance(self, z: _InputT, z2: _InputT | None = None, /) -> u.Quantity: ...
```
mypy interprets this as:
- The function always accepts one mandatory argument `z: _InputT`.
- Optionally accepts a second positional argument `z2: _InputT | None`.
- The return type is always u.Quantity.

This type-checks fine. The caller can do: 
```python
cosmo.angular_diameter_distance(0.5)
cosmo.angular_diameter_distance(0.5, 1.0)
```

What the `@overload` does is it says you shouldn't do
```python
cosmo.angular_diameter_distance(0.5, None)
# Mypy error: None not allowed for 2-argument overload
```

Now that I'm looking into this, maybe I should remove the `@overload`! `None` isn't pretty — it's better to omit the argument —  but it is valid.


> z1 and z2 are just poor choices of names

But they are positional-only parameters. Their names are meaningless. Which appears to be an advantage, since `z_lens` and `z_src` are use-case specific. `z_ref` is fine, but it's not that semantically meaningful, less so than it's argument position. 


Edit: I removed the overloads!

File: astropy/cosmology/_src/flrw/base.py [hunk 2]
```diff
@@ -809,10 +809,8 @@ def _integral_age(self, z: u.Quantity | ArrayLike, /) -> FArray:
 
     @overload
     def comoving_distance(self, z: _InputT, /) -> u.Quantity: ...
```
Comment (nstarman) [inline astropy/cosmology/_src/flrw/base.py:811]: I can remove the overload in this or a followup PR.

File: astropy/cosmology/_src/flrw/base.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/cosmology/_src/flrw/base.py:978]: > these aren't compatible with most jit operations

first time I'm hearing about this. Is this specific to a particular JIT compiler, or is it general ? if not, what JIT do you have in mind ?

About making the signature tighter, have you considered having just *one* positional-only argument, that would accept either a single Quantity or a *pair* of quantities ? I find this kind of pattern pretty intuitive to use, and it got so much easier to implement and maintain with structural pattern matching (`match/case`) that I think they are an attractive idea now.

Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:978]: > first time I'm hearing about this. Is this specific to a particular JIT compiler, or is it general ? if not, what JIT do you have in mind ?

Oh, thanks for catching that. I meant mapping operators — https://docs.pytorch.org/docs/stable/generated/torch.vmap.html, https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html.
For performance reasons these can also JIT compile.


> About making the signature tighter, have you considered having just one positional-only argument, that would accept either a single Quantity or a pair of quantities

I like this idea in principle.  The major drawback I see is how would you distinguish between the following?

```python
# Compute many distances
many_zs = (z0, z1, z2, ..., zN)  # shape (N,)
dists = Planck18.angular_diameter_distance(many_zs)  # shape (N,)

# Actually I just need 2
couple_zs = many_zs[:2]
dists = Planck18.angular_diameter_distance(couple_zs)  # shape (2,)
```

and 

```python
# Compute distance from z1 to z2
dist = Planck18.angular_diameter_distance((z1, z2))  # shape (1,)
```

With a pair of quantities it is impossible to guarantee full support for passing ArrayLike objects to `angular_diameter_distance`. With 2 arguments this is guaranteed.

```python
# Compute distance from z1 to z2
# The structure of the function means ArrayLike z1, z2 are always supported.
dist = Planck18.angular_diameter_distance(z1, z2)  # shape (1,)
```

Comment (neutrinoceros) [reply astropy/cosmology/_src/flrw/base.py:978]: Hum I'll admit I didn't think of that, *but* it *does* seem like a 2 element 1D array doesn't actually match the structure you'd write for a 2-element tuple
```py
>>> a = np.array((1, 2))
>>> match a:
...     case (_, _):
...         print('pair')
...     case _:
...         print('default')
...
default
```
is this a peculiarity in numpy, or would that be general to all Array API implementers ? if the latter, it looks to me that my idea still works as designed.

Comment (neutrinoceros) [reply astropy/cosmology/_src/flrw/base.py:978]: (btw, starting with keyword-only arguments and *maybe later* allow them to be passed positionally too would be backwards compatible, so we don't need to allow positional use from the get go to get the benefit of JIT compilation when we're ready to support it)

Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:978]: >  but it does seem like a 2 element 1D array doesn't actually match the structure you'd write for a 2-element tuple

But we want to support tuples as valid input as they are ArrayLike. So while 2-elt arrays may be distinguishable (though I have my doubts in general because of e.g. batch axes), the tuple behaviour makes this a non-starter.

Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:978]: > starting with keyword-only arguments and maybe later allow them to be passed positionally too would be backwards compatible.

I'm just not sure how this is an improvement. I have yet to see a semantically meaningful variable name I actually like, while the argument position follows standard mathematical convention and thus clearly conveys its meaning. Any confusion should be alleviated by the docstring (if it isn't then we should improve the docstrings).
And keyword arguments are incompatible with mapping ops ("Oh, thanks for catching that. I meant mapping operators" — https://github.com/astropy/astropy/pull/18807#discussion_r2547055652) so IMO it's best to be forward-thinking about this. 
A Positional-Or-Keyword argument (so moving the "/") would be fine by me, but only if we had a *good* semantically meaningful variable name. @mhvk's suggestion of `z_ref` is definitely the best I've heard so far and IMO it's only meh (is it a reference redshift or the redshift of reference / observer or observee?). 
The advantage of starting with Positional-Only is that it's backwards compatible with making an argument Positional-Or-Keyword. So IMO we should start with that and if ever there's a good variable name think about ~re~naming `z2` (it's positional and thus has no true name) and making it Positional-Or-Keyword.

File: astropy/cosmology/_src/flrw/base.py [hunk 3]
```diff
@@ -991,8 +1001,24 @@ def angular_diameter_distance(self, z: u.Quantity | ArrayLike, /) -> u.Quantity:
         .. [2] Weedman, D. (1986). Quasar astronomy, pp 65-67.
         .. [3] Peebles, P. (1993). Principles of Physical Cosmology, pp 325-327.
         """
-        z = aszarr(z)
-        return self.comoving_transverse_distance(z) / (z + 1.0)
+        z1, z2 = (0.0, z) if z2 is None else (z, z2)
+        z1, z2 = aszarr(z1), aszarr(z2)
+        if np.any(z2 < z1):
```
Comment (nstarman) [inline astropy/cosmology/_src/flrw/base.py:1006]: I'm thinking about removing this warning in the future. But since this function replaces the deprecated `angular_diameter_distance_z1z2` and that had the warning I think it's good to keep for now.

File: astropy/cosmology/_src/flrw/base.py (hunk 1 above)
Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:978]: @mhvk if you're interested where the overloads originally came from, it's https://cosmology.readthedocs.io/projects/api/latest/api/reference/cosmology.api.HasAngularDiameterDistance.angular_diameter_distance.html#cosmology.api.HasAngularDiameterDistance.angular_diameter_distance,
where we customized Sphinx to read the overloads to build more informative docs.
In that API document we've thought through improvements to the Astropy Cosmology API and I've been slowly working to transition us towards these improvements.

**Comment by nstarman** [comment]
This PR is part of the build-up towards [DistanceMeasures traits](#18808) and then [matter-only pedagogical cosmology]( #18876)

**Comment by nstarman** [comment]
@mhvk @neutrinoceros how do we move forward with this?  
https://github.com/astropy/astropy/pull/18807#discussion_r2547310503  indicates we need 2 arguments
https://github.com/astropy/astropy/pull/18807#discussion_r2547333153 walks through the problems with kwargs and how pos-only is more future-and-backwards compatible.

**Comment by nstarman** [comment]
@neutrinoceros I want to return to this before v8.0 feature freeze. 
A 2-elt tuple breaks array-like input and positional arguments matches the mathematical semantics with which cosmologists are familiar.

**Comment by nstarman** [comment]
Ping @neutrinoceros for a review, following our chat the other day.

**Comment by nstarman** [comment]
@neutrinoceros. Ping for a review, Thanks!

**Comment by neutrinoceros** [comment]
Sorry I got entrenched with my own gigantic PR (#19575). I'm doing reviews all day today !

File: astropy/cosmology/_src/flrw/base.py [hunk 4]
```diff
@@ -991,8 +1001,24 @@ def angular_diameter_distance(self, z: u.Quantity | ArrayLike, /) -> u.Quantity:
         .. [2] Weedman, D. (1986). Quasar astronomy, pp 65-67.
         .. [3] Peebles, P. (1993). Principles of Physical Cosmology, pp 325-327.
         """
-        z = aszarr(z)
-        return self.comoving_transverse_distance(z) / (z + 1.0)
+        z1, z2 = (0.0, z) if z2 is None else (z, z2)
+        z1, z2 = aszarr(z1), aszarr(z2)
```
Comment (neutrinoceros) [inline astropy/cosmology/_src/flrw/base.py:1005]: purely stylistic nit, feel free to ignore, though IMO this is easier to understand at a glance
```suggestion
        match z2:
            case None:
                z1 = aszarr(0.0)
                z2 = aszarr(z)
            case _:
                z1 = aszarr(z)
```

File: astropy/cosmology/_src/flrw/base.py (hunk 3 above)
Comment (neutrinoceros) [reply astropy/cosmology/_src/flrw/base.py:1006]: curious, what is the actual issue with z1<z2 ?

**Review by neutrinoceros — APPROVED** [review]
I took a good 20min to make sure I understood what I was signing off and I want to lift my previous concerns (bike shed ?) with the signature you elected. As you explaining, allowing these arguments to be passed by name only makes sense if we have good names for them, which so far we don't, so, I'm planning to merge this in a couple days if no objection is raised in the mean time.

**Review by neutrinoceros — COMMENTED** [review]
RTD failure is unrelated and fixed on main already (https://github.com/astropy/astropy/pull/19594). Unfortunately they still hide any real failure that would be caused by changes from this PR, so I made a couple suggestions where I had doubts (some of which I suspect may not be needed) to increase our chances that the next build will indeed succeed.

File: astropy/cosmology/_src/flrw/base.py (hunk 4 above)
Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:1005]: Given that this will appear in a few methods, I think in a followup PR these lines should be shunted off to a helper function. I'd want to do timing tests then. Also, I think this helper function + aszarr are good targets for C code.
Given that, I'm going to push off rewrites here.

File: astropy/cosmology/_src/flrw/base.py (hunk 3 above)
Comment (nstarman) [reply astropy/cosmology/_src/flrw/base.py:1006]: z2<z1 makes the distance go negative.

**Comment by nstarman** [comment]
Thanks @neutrinoceros for the review!

**Comment by pllim** [comment]
I opened https://github.com/astropy/astropy/issues/19613 as future follow-up issue. Thanks, all!
